trin-core/src/types/block_body.rs

```diff
@@ -0,0 +1,307 @@
+        let mut stream = RlpStream::new();
+        match self {
+            Self::Legacy(tx) => {
+                tx.rlp_append(&mut stream);
+                stream.out().freeze().to_vec()
+            }
+            Self::AccessList(tx) => {
+                tx.rlp_append(&mut stream);
+                [&[TransactionId::AccessList as u8], stream.as_raw()].concat()
+            }
+            Self::EIP1559(tx) => {
+                tx.rlp_append(&mut stream);
+                [&[TransactionId::EIP1559 as u8], stream.as_raw()].concat()
+            }
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, RlpEncodable, RlpDecodable)]
+pub struct LegacyTransaction {
+    pub nonce: U256,
+    pub gas_price: U256,
+    pub gas: U256,
+    pub to: Address,
+    pub value: U256,
+    pub data: Bytes,
+    pub v: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Eq, Debug, Clone, PartialEq, RlpDecodable, RlpEncodable)]
+pub struct AccessListTransaction {
+    pub chain_id: U256,
+    pub nonce: U256,
+    pub gas_price: U256,
+    pub gas_limit: U256,
+    pub to: H160,
+    pub value: U256,
+    pub data: Bytes,
+    pub access_list: AccessList,
+    pub y_parity: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Eq, Debug, Clone, PartialEq, RlpDecodable, RlpEncodable)]
+pub struct EIP1559Transaction {
+    pub chain_id: U256,
+    pub nonce: U256,
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
+    pub gas_limit: U256,
+    pub to: H160,
+    pub value: U256,
+    pub data: Bytes,
+    pub access_list: AccessList,
+    pub y_parity: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct AccessList {
+    pub list: Vec<AccessListItem>,
+}
+
+impl Decodable for AccessList {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        let list = rlp
+            .iter()
+            .map(|v| rlp::decode(v.as_raw()).unwrap())
+            .collect();
+        Ok(Self { list })
+    }
+}
+
+impl Encodable for AccessList {
+    fn rlp_append(&self, stream: &mut RlpStream) {
+        stream.begin_list(self.list.len());
+        for access in self.list.iter() {
+            stream.append(access);
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Eq)]
+pub struct AccessListItem {
+    pub addr: H160,
+    pub keys: StorageKeys,
+}
+
+impl Decodable for AccessListItem {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        let addr = rlp.val_at(0)?;
+        let keys: Vec<H256> = rlp.list_at(1)?;
+        Ok(Self {
+            addr,
+            keys: StorageKeys { keys },
+        })
+    }
+}
+
+impl Encodable for AccessListItem {
+    fn rlp_append(&self, stream: &mut RlpStream) {
+        stream.begin_list(2);
+        stream.append(&self.addr);
+        stream.begin_list(self.keys.keys.len());
+        for location in &self.keys.keys {
+            stream.append(location);
+        }
```

Comment:
`keys.keys` is surprising here. Maybe the `AccessListItem` variable could be `storage`. If `append_list` works, then this could be:
```suggestion
        stream.append_list(self.storage.keys);
```